platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/http/OkHttpRequest.java

```diff
@@ -0,0 +1,256 @@
+        log(Log.VERBOSE, String.format("[HTTP] Request was successful (code = %s).", response.code()));
+      } else {
+        // We don't want to call this unsuccessful because a 304 isn't really an error
+        String message = !TextUtils.isEmpty(response.message()) ? response.message() : "No additional information";
+        log(Log.DEBUG, String.format("[HTTP] Request with response code = %s: %s", response.code(), message));
+      }
+
+      removeCall(call);
+
+      ResponseBody responseBody = response.body();
+      if (responseBody == null) {
+        log(Log.ERROR, "[HTTP] Received empty response body");
+        return;
+      }
+
+      byte[] body;
+      try {
+        body = responseBody.bytes();
+      } catch (IOException ioException) {
+        onFailure(call, ioException);
+        // throw ioException;
+        return;
+      } finally {
+        response.close();
+      }
+
+      httpRequest.onResponse(response.code(),
+        response.header("ETag"),
+        response.header("Last-Modified"),
+        response.header("Cache-Control"),
+        response.header("Expires"),
+        response.header("Retry-After"),
+        response.header("x-rate-limit-reset"),
+        body);
+    }
+
+    private void handleFailure(Call call, Exception e) {
+      String errorMessage = e.getMessage() != null ? e.getMessage() : "Error processing the request";
+      int type = getFailureType(e);
+
+      if (logEnabled && call != null && call.request() != null) {
+        String requestUrl = call.request().url().toString();
+        logFailure(type, errorMessage, requestUrl);
+      }
+
+      removeCall(call);
+      Timber.e("Handle Failure with size %s for %s", calls.size(), call.hashCode());
+      httpRequest.handleFailure(type, errorMessage);
+    }
+
+    private void logFailure(int type, String errorMessage, String requestUrl) {
+      log(type == TEMPORARY_ERROR ? Log.DEBUG : type == CONNECTION_ERROR ? Log.INFO : Log.WARN,
+        String.format(
+          "Request failed due to a %s error: %s %s",
+          type == TEMPORARY_ERROR ? "temporary" : type == CONNECTION_ERROR ? "connection" : "permanent",
+          errorMessage,
+          logRequestUrl ? requestUrl : ""
+        )
+      );
+    }
+
+    private int getFailureType(Exception e) {
+      if ((e instanceof NoRouteToHostException) || (e instanceof UnknownHostException) || (e instanceof SocketException)
+        || (e instanceof ProtocolException) || (e instanceof SSLException)) {
+        return CONNECTION_ERROR;
+      } else if ((e instanceof InterruptedIOException)) {
+        return TEMPORARY_ERROR;
+      }
+      return PERMANENT_ERROR;
+    }
+
+    private void removeCall(Call call) {
+      synchronized (calls) {
+        Call currentCall;
+        for (int i = 0; i < calls.size(); i++) {
+          long key = calls.keyAt(i);
+          currentCall = calls.get(key);
+          if (call.equals(currentCall)) {
+            calls.delete(key);
+          }
```

Comment:
I believe we can safely return here.

Reply:
yes, also updated code block to:
Which should also be faster as it avoids using `get(key)`
```java
      synchronized (calls) {
        Call currentCall;
        for (int i = 0; i < calls.size(); i++) {
          currentCall = calls.valueAt(i);
          if (call.equals(currentCall)) {
            calls.delete(calls.keyAt(i));
            return;
          }
        }
      }
```